### ISSUE TITLE
Support SQLcipher

### ISSUE THREAD

**Issue by talmeme** [vapor/sqlite-nio#72]
[SQLcipher](https://github.com/sqlcipher/sqlcipher) is a standalone fork of the [SQLite](https://www.sqlite.org/) database library that adds 256 bit AES encryption of database files and other security features like:

- on-the-fly encryption
- tamper detection
- memory sanitization
- strong key derivation

SQLcipher uses one of CommonCrypto, libtomcrypt, NSS and OpenSSL for crypto functionality.

**Comment by talmeme** [comment]
I have a [working integration](https://github.com/talmeme/sqlite-nio/tree/sqlcipher), with corresponding [SQLite-Kit](https://github.com/talmeme/sqlite-kit/tree/sqlcipher) and [Fluent SQLite driver](https://github.com/talmeme/fluent-sqlite-driver/tree/sqlcipher).

And a [demo](https://github.com/talmeme/vapor-sqlcipher-demo). Tested working on macOS Sonoma aarch64 and Ubuntu 22.04 aarch64.

**Comment by 0xTim** [comment]
If this can be achieved using Swift Crypto we can probably looking at merging it. Otherwise it's probably a no go

**Comment by talmeme** [comment]
> If this can be achieved using Swift Crypto we can probably looking at merging it. Otherwise it's probably a no go

SQLcipher is implemented in C using SQLite's deprecated `codec` hooks and compiled together with SQLite's C source. This likely precludes using Swift Crypto to implement an SQLcipher crypto provider.

An alternative approach is to use SQLite's [VFS mechanism](https://sqlite.org/vfs.html), as is done by [SQLite3MultipleCiphers](https://github.com/utelle/SQLite3MultipleCiphers). Not sure how feasible to implement an SQLite VFS in Swift.

**Comment by 0xTim** [comment]
I'll let Gwynne chime in when she's back as she knows this far better than I do. The main issue is the inclusion of a system dependency on tomcrypt which complicates a number of things, including the fact that it hasn't had a release in 6 years and I find it very hard to believe there haven't been any security issues in that time.

OpenSSL is probably the only possibility if we need to link against something, but again, its a bulky library and another dependency that's required regardless of whether you're using it or not

**Comment by talmeme** [comment]
I agree on tomcrypt. I use it because that's what [Skip](https://github.com/skiptools/skip-sql) does, so I started with something known working.

I now see that it is easy to link SQLcipher with OpenSSL. BoringSSL works too, although, as you have mentioned, BoringSSL doesn't have stable API and usually needs to be built from source.

**Comment by talmeme** [comment]
Updates:
- The tomcrypt version is moved to https://github.com/talmeme/sqlite-nio/tree/sqlcipher-ltc branch.
- The https://github.com/talmeme/sqlite-nio/tree/sqlcipher branch now builds with OpenSSL, tested on Ubuntu 22.04 x86_64, aarch64, and macOS Sonoma aarch64 (with macports).

**Comment by dioKaratzas** [comment]
@talmeme any update on this one? Can you please rebase the repos? wanna give it a try on an iOS app currently building

**Comment by talmeme** [comment]
> [@talmeme](https://github.com/talmeme) any update on this one? Can you please rebase the repos? wanna give it a try on an iOS app currently building

I did this work to [SQLcipher-enable litestream](https://github.com/talmeme/litestream/tree/sqlcipher). I have a private Vapor app running on a VPS that does encrypted litestream streaming to S3-compatible storage. At the moment I can't tell when I'll pick this up again. 

For local SQLite encryption in an iOS app there should be other options.

**Comment by gwynne** [comment]
At this time, I don't see a path forward for adding this. If you need this, I would recommend using it in a fork.
